ui/server.py

```diff
@@ -65,38 +92,152 @@ class ImageRequest(BaseModel):
+class TaskCache():
+    def __init__(self):
+        self._base = dict()
+    def _get_ttl_time(self, ttl: int) -> int:
+        return int(time.time()) + ttl
+    def _is_expired(self, timestamp: int) -> bool:
+        return int(time.time()) >= timestamp
+    def clean(self) -> None:
+        for key in self._base:
+            ttl, _ = self._base[key]
+            if self._is_expired(ttl):
+                del self._base[key]
+    def clear(self) -> None:
+        self._base.clear()
+    def delete(self, key: Hashable) -> bool:
+        if key not in self._base:
+            return False
+        del self._base[key]
+        return True
+    def keep(self, key: Hashable, ttl: int) -> bool:
+        if key in self._base:
+            _, value = self._base.get(key)
+            self._base[key] = (self._get_ttl_time(ttl), value)
+            return True
+        return False
+    def put(self, key: Hashable, value: Any, ttl: int) -> bool:
+        try:
+            self._base[key] = (
+                self._get_ttl_time(ttl), value
+            )
+        except Exception:
+            return False
+        return True
+    def tryGet(self, key: Hashable) -> Any:
+        ttl, value = self._base.get(key, (None, None))
+        if ttl is not None and self._is_expired(ttl):
+            self.delete(key)
+            return None
+        return value
+
+task_cache = TaskCache()
+
 class SetAppConfigRequest(BaseModel):
     update_branch: str = "main"
 
 @app.get('/')
 def read_root():
-    headers = {"Cache-Control": "no-cache, no-store, must-revalidate", "Pragma": "no-cache", "Expires": "0"}
-    return FileResponse(os.path.join(SD_UI_DIR, 'index.html'), headers=headers)
-
-@app.get('/ping')
-async def ping():
-    global model_loaded, model_is_loading
-
+    return FileResponse(os.path.join(SD_UI_DIR, 'index.html'), headers=NOCACHE_HEADERS)
+
+def preload_model(file_path=None):
+    global current_state, current_state_error, current_model_path
+    if file_path == None:
+        file_path = get_initial_model_to_load()
+    if file_path == current_model_path:
+        return
+    current_state = ServerStates.LoadingModel
     try:
-        if model_loaded:
-            return {'OK'}
-
-        if model_is_loading:
-            return {'ERROR'}
-
-        model_is_loading = True
-
         from sd_internal import runtime
-
-        runtime.load_model_ckpt(ckpt_to_use=get_initial_model_to_load())
-
-        model_loaded = True
-        model_is_loading = False
-
-        return {'OK'}
+        runtime.load_model_ckpt(ckpt_to_use=file_path)
+        current_model_path = file_path
+        current_state_error = None
+        current_state = ServerStates.Online
     except Exception as e:
+        current_model_path = None
+        current_state_error = e
+        current_state = ServerStates.Unavailable
         print(traceback.format_exc())
-        return HTTPException(status_code=500, detail=str(e))
+
+def thread_render():
```

Comment:
Can we move all the task-related classes and functions to another file? Maybe something like `task_manager.py`? So we'd have 3 files: `server.py` (handles the server interface), `task_manager.py` (manages the tasks and queue), `runtime.py` (interacts with the underlying SD engine. Maybe we should call it `renderer.py`?)

Reply:
We could for sure, but I didn't want to add one more file to distribute.
At this point it's just cleaning up a little and should be easy now that it's working.

Reply:
Adding another file is not a problem. The auto-updater copies the entire UI folder over, so we don't need to do anything extra wrt that.

So this would mean:
1. add `sd_internal/task_manager.py`
2. in `server.py`, use `from sd_internal import task_manager`, and update the references in the code to `task_manager.foo_function()` instead of `foo_function()`
3. in `task_manager.py` use `import .runtime` incase you need to interact with the runtime.

I'd definitely definitely prefer to keep these separate, and move the task-related code into `task_manager.py`. I'm trying to avoid adding additional bloat to already-bloated files, and am working on gradually refactoring the code (JS and Python). I gave a similar comment to the JS-related PR submitted earlier today by another developer.

Thanks!